### PR TITLE
Use problem globs to declare problem storage roots

### LINKED ISSUE #723: Watchdog fails if `problem_storage_root` is a string
Should check if `problem_dirs` is a string before https://github.com/DMOJ/judge-server/blob/master/dmoj/judgeenv.py#L173, otherwise we will be watching nonsense paths,

### LINKED ISSUE #772: Use globs for `problem_storage_root`
Declare problem storage roots using globs, replacing all 3 previous implementations.

For judge v2, this should be called `problem_storage_globs`, for compatibility.

For judge v3, this should replace the current implementation of `problem_storage_root`.

### PR DESCRIPTION
**Pull request by kiritofeng**
Closes #772. Also supersedes #1026.

### CODE REVIEW

**Comment by Xyene** [comment]
This largely looks good to me. Could you update the PR description to include a sample config?

I'd like the next version we tag to be 3.0.0, since we have a fair number of breaking changes. So, I think we should:

* merge this PR
* queue up a PR totally removing the old config format, but don't merge it until right before tagging 3.0.0 (so we don't wreck people installing from master)
* queue up a PR updating the docs

**Comment by kiritofeng** [comment]
As an example,
```yaml
problem_storage_root:
- 2: /problems/ccc/nccc
- 1: /problems/apio
- 1: /problems/bbc
```
would become
```yaml
problem_storage_globs:
- /problems/ccc/nccc/*/**/
- /problems/apio/*/apio*/
- /problems/bbc/*/bbc*/
```

File: dmoj/judgeenv.py
Comment (kiritofeng) [inline dmoj/judgeenv.py:230]: Something we could do (when we're ready to replace `problem_storage_root` with `problem_storage_globs`) is to replace the test regexes with test globs

File: dmoj/judgeenv.py
Comment (Xyene) [inline dmoj/judgeenv.py:203]:
```suggestion
        startup_warnings.append('`problem_storage_root` is deprecated, use `problem_storage_globs` instead')
```

File: dmoj/judgeenv.py
Comment (Xyene) [reply dmoj/judgeenv.py:229]: This sounds like a fine idea to leave as a comment :)

File: dmoj/judgeenv.py
Comment (Xyene) [inline dmoj/judgeenv.py:197]: We should not support this mode, and instead always require a list here.

File: dmoj/judgeenv.py
Comment (Xyene) [inline dmoj/judgeenv.py:352]: When can this happen?

Comment (kiritofeng) [reply dmoj/judgeenv.py:352]: if you delete a folder that previously contained a problem, it can cause the scan to crash

Comment (Xyene) [reply dmoj/judgeenv.py:352]: Let's invert this to read `if not ...: continue` and leave a comment.

Is this new behavior, or did it always crash this way and this is fixing a bug?

File: dmoj/judgeenv.py
Comment (Xyene) [inline dmoj/judgeenv.py:224]:
```suggestion
            raise SystemExit('`problem_storage_globs` not specified in "%s"; no problems available to grade' % model_file)
```

File: dmoj/judgeenv.py
Comment (kiritofeng) [reply dmoj/judgeenv.py:352]: Just tested on the judges we run in prod, the bug exists there too

**Review by Xyene — APPROVED** [review]
LGTM, feel free to merge once lint passes.

**Pull request merged**
